Auto-Expand sub-folder when dragging over it. Issue #492

nhold opened the pull request:
### Auto-Expand for TreeViewAdv. Issue #492.

I went with the approach to extend TreeViewAdv and used the new class in both ProjectFolderView and SceneView. I tried to keep the changes as minimal as possible although the solution and csproj seem to have run away from a simple add existing file, let me know if I should remove them.

A few questions\comments:

1. Do you want to keep consistency of the type held by the SceneView\ProjectFolderView? I have currently kept the type as TreeViewAdv and just newed it to the new type (AutoExpandTreeView) which has a default of 1000 ms delay.

2. Is there a place to provide user\editor preferences, I would suggest putting the delay as an option for users. 

3. Have fun reviewing <3

4. Should I add automated testing?

Manual Tests:

![test](https://cloud.githubusercontent.com/assets/1107474/23647088/b61b3ada-035f-11e7-98d3-693d40b98b5d.gif)
![test2](https://cloud.githubusercontent.com/assets/1107474/23647090/b7f3c9ee-035f-11e7-81c8-a97071e419e4.gif)

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+{
+	public class AutoExpandTreeView : TreeViewAdv
+	{
+		public int DragOverExpandDelay
```
We generally don't do auto-properties in Duality source. Please replace with a regular field-property combo.

nhold replied:
Looking at this a bit more, it's really just a property exposing the the Interval property of the timer. Do code guidelines prefer just doing the following:

```
private Timer       autoExpandDelayTimer = new Timer();
public int AutoExpandDelay
{
    get
    {
        return autoExpandDelayTimer.Interval;
    }

    set
    {
        autoExpandDelayTimer.Interval = value;
    }
}
```

ilexp replied:
Looks good semantically! Style-wise, for property implementations that short, the single-line form is preferred:

```CSharp
private Timer autoExpandDelayTimer = new Timer();

public int AutoExpandDelay
{
    get { return this.autoExpandDelayTimer.Interval; }
    set { this.autoExpandDelayTimer.Interval = value; }
}
```

In other cases, we just omit the newline between getter and setter.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+			set;
+		} = 1000;
+
+		private TreeNodeAdv autoExpandNode = null;
```
Code Style: All fields as a group, no empty lines between them except to signify larger groups of different categories of fields (none in this case). Also, fields are the first thing in a class definition, followed by properties, and then methods.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+
+		private TreeNodeAdv autoExpandNode = null;
+
+		private Timer timer = new Timer();
```
"timer" is very general! This is specifically an auto-expand timer. Suggesting to rename this field. Same goes for the Start / Stop methods.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+
+		private Timer timer = new Timer();
+
+		private void StartTimer()
```
See above

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+			}
+		}
+
+		private void StopTimer()
```
See above

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+			{
+				timer.Stop();
+				timer.Interval = DragOverExpandDelay;
+				timer.Tick += OnTimerTick;
```
You never seem to unsubscribe the event handler again - could be a bug. Please verify. Suggesting to unsubscribe in the `StopTimer` method.

ilexp replied:
On second thought, there is no reason to subscribe / unsubscribe that often at all. You could just subscribe the event in the constructor and unsubscribe in a `Dispose` override and just start / stop the timer as needed. A positive side effect would be to produce less garbage due to implicit delegate instantiations.

nhold replied:
Yep, we were re-listening every time, possibly degrading performance over time. Now just doing it in a constructor and unsubscribing in `Dispose`. Also added an optional parameter for the constructor to take in the ms for the `autoExpandDelayTimer` interval (1000 is default).

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+
+		private void StartTimer()
+		{
+			if (DragOverExpandDelay > 0)
```
Expanding on the above no-auto-properties comment: When there is no specific reason to do otherwise, please use the backing field directly inside your class. No need to invoke a getter method. This is for all occurrences of get-access of `DragOverExpandDelay`.

ilexp replied:
Something else: All member access (fields, properties, methods) needs to be preceded with a `this.` - this is a code convention in the Duality repository.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+
+			if (DragOverExpandDelay > 0)
+			{
+
```
No reason to have an empty line here. Please remove for clarity / consistency / readability.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+			StopTimer();
+		}
+
+		protected override void OnDragOver(DragEventArgs drgevent)
```
There's no reason to abbreviate `drgevent` that strongly (although that just might have been the default name). Suggesting to use `dragEvent` instead.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+			{
+
+				Point p = PointToClient(new Point(drgevent.X, drgevent.Y));
+				var node = GetNodeAt(p);
```
Another coding convention: Don't use `var` unless the alternative would decrease readability, such as `foreach`ing over a dictionary with complicated type parameters. This is not one of those cases, please no `var` here :) Also (super high level), suggesting to rename this to `hoveredNode`.

ilexp commented on Source/Editor/DualityEditor/Controls/AutoExpandTreeView.cs:
```diff
@@ -0,0 +1,68 @@
+			if (DragOverExpandDelay > 0)
+			{
+
+				Point p = PointToClient(new Point(drgevent.X, drgevent.Y));
```
Suggesting rename to `localHoverPos` or a similar descriptive name of your choice.

nhold commented:
> Something else: All member access (fields, properties, methods) needs to be preceded with a this. - this is a code convention in the Duality repository.

Does this include inherited members? I.e `this.PointToClient(...)`?

Other than the above I believe all the code review changes have gone in, plus the following bug fixes\changes:

1. You can no longer open children of the node you are dragging or their children (You can't move a parent to it's children) - If this is expected behaviour let me know.
2. Dragging a node onto a valid 'auto expand' node and then off of the local client area no longer continues to expand the node you last dragged over.
3. If you have already dropped a node the previous node you dragged over will no longer expand afterwards.
4. Fixed crash when moving from project->scene view

There is one bug left to go when dragging project->scene view and vice versa. The behaviour is inconsistent on whether it will expand or not. Not sure if this is something I introduced but when drag and dropping a prefab from project to scene it is created with an empty name.

ilexp commented:
> > Something else: All member access (fields, properties, methods) needs to be preceded with a this. - this is a code convention in the Duality repository.
> 
> Does this include inherited members? I.e `this.PointToClient(...)`?

Yep! All of them.

> Other than the above I believe all the code review changes have gone in, plus the following bug fixes\changes:
> 
> 1. You can no longer open children of the node you are dragging or their children (You can't move a parent to it's children) - If this is expected behaviour let me know.
> 2. Dragging a node onto a valid 'auto expand' node and then off of the local client area no longer continues to expand the node you last dragged over.
> 3. If you have already dropped a node the previous node you dragged over will no longer expand afterwards.
> 4. Fixed crash when moving from project->scene view

Nice work 👍 

> There is one bug left to go when dragging project->scene view and vice versa. The behaviour is inconsistent on whether it will expand or not. Not sure if this is something I introduced but when drag and dropping a prefab from project to scene it is created with an empty name.

Hmm.. I wouldn't know how your changes should introduce that kind of Prefab dragdrop bug, but that's definitely not normal behavior - and I can't reproduce it in an unchanged build. This is how it should look like: https://gfycat.com/KaleidoscopicQuarrelsomeEarthworm

Can you verify whether this is a regression introduced by this PR? _Maybe_ you are somehow accidentally interfering with something that may or may not trigger renaming-related code, leading to emptying the name of a newly instantiated node, but.. just wildly guessing here.

One more issue regarding code structure: Don't do empty lines between methods, except if the method above and the one below belong to different "groups" or "categories of methods. Which grouping you use is up for debate and your judgement, but there should be no empty line without reason. (Collapsed) example:

```CSharp
public AutoExpandTreeView(int autoExpandDelay = 1000)

private void StartAutoExpandTimer()
private void StopAutoExpandTimer()

private void OnAutoExpandTimerTick(object sender, EventArgs e)
protected override void OnDragOver(DragEventArgs dragEvent)
protected override void OnDragLeave(EventArgs dragLeaveEvent)
protected override void OnDragDrop(DragEventArgs dragDropEvent)

protected override void Dispose(bool disposing)
```

I'll let you address the last few issues, but I think we're making good progress here.

> P.s Let me know if I should squash all these commits.

No need to, I'll use the GitHub Squash-Merge feature on this PR.

nhold commented:
Okay that bug seems to be on master as well (on commit e94e8ca) if you create an empty prefab and drag it into the scene it creates a GameObject with an empty string as a name. Not sure if it's expected behaviour, here it is:

![test3](https://cloud.githubusercontent.com/assets/1107474/23750901/bed4d494-051a-11e7-809d-c5676484e336.gif)

Latest commit employs recursion to go through parents of the hoveredNode to ensure the we aren't dragging a parent over it. I added an extension method for that functionality.

ilexp commented:
> Okay that bug seems to be on master as well (on commit e94e8ca) if you create an empty prefab and drag it into the scene it creates a GameObject with an empty string as a name. Not sure if it's expected behaviour, here it is:

Ah - good find. I didn't test an empty one. Prefabs remain empty until an existing object is injected into them, but usually they're just created by dragging an object into the Project View. The behavior in your use case is expected, but admittedly not intuitive. Will have to figure out how to deal with this case better.

> On top of that I believe all the stylistic changes are addressed.

They are. Thanks for your contribution and especially for being so positive about change requests! I'll merge this and you will find this update in the next binary release.
